### DISCUSSION TITLE
Custom plugin persist to database

### DISCUSSION THREAD

**Discussion by wolflu05** [inventree/InvenTree#3523] [Q&A] [answered]
Hello,

I'm going forward with my [plugin](https://github.com/wolflu05/inventree-bulk-plugin). Now I want to implement saving of bulk plugin schemas to the database as templates for later usage. Now my question is, how can I do this? Ideally I'd like to have a custom form for inputing a schema for the bulk creation, as I later plan to replace the textarea for the yml input with some more advanced editor like Monaco or a json visual editor.

**Comment by wolflu05** [comment, marked as answer]
I think I found the `AppMixin` which is suitable for custom models. I took a look at @matmair 's plugins ([ShopifyIntegrationPlugin](https://github.com/matmair/ShopifyIntegrationPlugin) and [inventree-zapier](https://github.com/matmair/inventree-zapier)). One tricky part is the `NAME` must be equal to the folder in which the main module lives and not the filename, otherwise we get an `missing AppConfig mixin` error from Django.

#### But I still have one question left

How do migrations work? I created a folder `migrations` like its in the ShopifyPlugin, but if I invoke `migrate` the migrations won't be applied. You can take a look at my current files at [my repo](https://github.com/wolflu05/inventree-bulk-plugin/tree/dev).

### EDIT
see #3549 for reference

**Reply by SchrodingersGat** [reply]
@matmair this one might be one for you :)

**Reply by wolflu05** [reply]
@SchrodingersGat I now added a \_\_init\_\_ file to my migrations folder but it doesn’t seem to be recognized by the migration runner. So I added a few debug points in the registry.py file and saw that the plugin loading mechanism is not triggered for the migrations. So I guess this plugin app is not installed to Django during migrations.

**Reply by matmair** [reply]
@wolflu05 migrations are fed with the normal list of apps - including app plugins if enabled - for generation. I am a bit short on time rn but I will take a look once I find a slot for it.

**Reply by matmair** [reply]
@wolflu05 I checked out the issue and the behaviour for packages is hard-coded right now.
That has no technical reason - it just reflects that I build plugins around my workflows - in which `NAME` and package-name are atomically the same.

We can only change it in `0.9.9` as it will be breaking but I think we can change it till then. There are at least a few changes lined up for plugins so it does not hurt.

**Reply by wolflu05** [reply]
@matmair Thank you for your rare time. Do you also found something out regarding the migrations?

**Reply by wolflu05** [reply]
Seems like `inv server` discovers that there is an unapplied migration but `inv migrate` don't.

<details>

```bash
vscode ➜ /workspaces/InvenTree (master ✗) $ inv server
Waiting for database...
Database connection sucessful!
Watching for file changes with StatReloader
Watching for file changes with StatReloader
Performing system checks...

System check identified no issues (1 silenced).

You have 1 unapplied migration(s). Your project may not work properly until you apply the migrations for app(s): inventree_bulk.
Run 'python manage.py migrate' to apply them.
August 14, 2022 - 07:40:01
Django version 3.2.15, using settings 'InvenTree.settings'
Starting development server at http://127.0.0.1:8000/
Quit the server with CONTROL-C.
^Cvscode ➜ /workspaces/InvenTree (master ✗) $ inv migrate
Running InvenTree database migrations...
========================================
No changes detected
Operations to perform:
  Apply all migrations: account, admin, auth, authtoken, build, common, company, contenttypes, django_q, error_report, exchange, label, order, otp_static, otp_totp, part, plugin, report, sites, socialaccount, stock, user_sessions, users
Running migrations:
  No migrations to apply.
Operations to perform:
  Synchronize unmigrated apps: InvenTree, allauth, allauth_2fa, corsheaders, crispy_forms, django_cleanup, django_filters, django_otp, djmoney, formtools, import_export, maintenance_mode, markdownify, messages, mptt, rest_framework, sslserver, staticfiles
  Apply all migrations: account, admin, auth, authtoken, build, common, company, contenttypes, django_q, error_report, exchange, label, order, otp_static, otp_totp, part, plugin, report, sites, socialaccount, stock, user_sessions, users
Synchronizing apps without migrations:
  Creating tables...
    Running deferred SQL...
Running migrations:
  No migrations to apply.
System check identified no issues (1 silenced).
========================================
InvenTree database migrations completed!
Rebuilding Part objects
Rebuilding PartCategory objects
Rebuilding StockItem objects
Rebuilding StockLocation objects
Rebuilding Build objects
vscode ➜ /workspaces/InvenTree (master ✗) $
```

</details>
